routing/bicycle_directions.cpp

```diff
@@ -0,0 +1,216 @@
+    , m_pathSegments(pathSegments)
+    , m_routeLength(0)
+  {
+    for (auto const & edge : routeEdges)
+    {
+      m_routeLength += MercatorBounds::DistanceOnEarth(edge.GetStartJunction().GetPoint(),
+                                                       edge.GetEndJunction().GetPoint());
+    }
+  }
+
+  // turns::IRoutingResult overrides:
+  virtual TUnpackedPathSegments const & GetSegments() const override { return m_pathSegments; }
+
+  virtual void GetPossibleTurns(TNodeId node, m2::PointD const & ingoingPoint,
+                                m2::PointD const & junctionPoint, size_t & ingoingCount,
+                                TurnCandidates & outgoingTurns) const override
+  {
+    ingoingCount = 0;
+    outgoingTurns.candidates.clear();
+
+    auto const adjacentEdges = m_adjacentEdges.find(node);
+    if (adjacentEdges == m_adjacentEdges.cend())
+    {
+      ASSERT(false, ());
+      return;
+    }
+
+    ingoingCount = adjacentEdges->second.m_ingoingTurnsCount;
+    outgoingTurns.candidates = adjacentEdges->second.m_outgoingTurns.candidates;
+  }
+
+  virtual double GetPathLength() const override { return m_routeLength; }
+
+  virtual m2::PointD const & GetStartPoint() const override
+  {
+    CHECK(!m_routeEdges.empty(), ());
+    return m_routeEdges.front().GetStartJunction().GetPoint();
+  }
+
+  virtual m2::PointD const & GetEndPoint() const override
+  {
+    CHECK(!m_routeEdges.empty(), ());
+    return m_routeEdges.back().GetEndJunction().GetPoint();
+  }
+
+private:
+  IRoadGraph::TEdgeVector const & m_routeEdges;
+  TAdjacentEdgesMap const & m_adjacentEdges;
+  TUnpackedPathSegments const & m_pathSegments;
+  double m_routeLength;
+};
+}  // namespace
+
+namespace routing
+{
+BicycleDirectionsEngine::BicycleDirectionsEngine(Index const & index) : m_index(index) {}
+
+void BicycleDirectionsEngine::Generate(IRoadGraph const & graph, vector<Junction> const & path,
+                                       Route::TTimes & times, Route::TTurns & turns,
+                                       vector<m2::PointD> & routeGeometry,
+                                       my::Cancellable const & cancellable)
+{
+  size_t const pathSize = path.size();
+  CHECK_NOT_EQUAL(pathSize, 0, ());
+
+  times.clear();
+  turns.clear();
+  routeGeometry.clear();
+  m_adjacentEdges.clear();
+  m_pathSegments.clear();
+
+  auto emptyPathWorkaround = [&]()
+  {
+    turns.emplace_back(pathSize - 1, turns::TurnDirection::ReachedYourDestination);
+    this->m_adjacentEdges[0] = AdjacentEdges(1);  // There's one ingoing edge to the finish.
+  };
+
+  if (pathSize <= 1)
+  {
+    ASSERT(false, (pathSize));
+    emptyPathWorkaround();
+    return;
+  }
+
+  CalculateTimes(graph, path, times);
+
+  IRoadGraph::TEdgeVector routeEdges;
+  if (!ReconstructPath(graph, path, routeEdges, cancellable))
+  {
+    LOG(LDEBUG, ("Couldn't reconstruct path"));
+    emptyPathWorkaround();
+    return;
+  }
+  if (routeEdges.empty())
+  {
+    ASSERT(false, ());
```

Comment:
Why? The same, again - single point path without edges is not a problem. Please, write a test that builds a route from the point to itself.